src/maxflow.rs

```diff
@@ -1 +1,251 @@
+            flow_limit,
+            level: vec![0; n_],
+            iter: vec![0; n_],
+            que: SimpleQueue::default(),
+        };
+
+        let mut flow = Cap::zero();
+        while flow < flow_limit {
+            calc.bfs();
+            if calc.level[t] == -1 {
+                break;
+            }
+            calc.iter.iter_mut().for_each(|e| *e = 0);
+            while flow < flow_limit {
+                let f = calc.dfs(t, flow_limit - flow);
+                if f == Cap::zero() {
+                    break;
+                }
+                flow += f;
+            }
+        }
+        flow
+    }
+
+    pub fn min_cut(&self, s: usize) -> Vec<bool> {
+        let mut visited = vec![false; self._n];
+        let mut que = SimpleQueue::default();
+        que.push(s);
+        while !que.empty() {
+            let &p = que.front();
+            que.pop();
+            visited[p] = true;
+            for e in &self.g[p] {
+                if e.cap != Cap::zero() && !visited[e.to] {
+                    visited[e.to] = true;
+                    que.push(e.to);
+                }
+            }
+        }
+        visited
+    }
+}
+
+struct FlowCalculator<'a, Cap> {
+    graph: &'a mut MfGraph<Cap>,
+    s: usize,
+    t: usize,
+    flow_limit: Cap,
+    level: Vec<i32>,
+    iter: Vec<usize>,
+    que: SimpleQueue<usize>,
+}
+
+impl<Cap> FlowCalculator<'_, Cap>
+where
+    Cap: MfCapacity,
+{
+    fn bfs(&mut self) {
+        self.level.iter_mut().for_each(|e| *e = -1);
+        self.level[self.s] = 0;
+        self.que.clear();
+        self.que.push(self.s);
+        while !self.que.empty() {
+            let v = *self.que.front();
+            self.que.pop();
+            for e in &self.graph.g[v] {
+                if e.cap == Cap::zero() || self.level[e.to] >= 0 {
+                    continue;
+                }
+                self.level[e.to] = self.level[v] + 1;
+                if e.to == self.t {
+                    return;
+                }
+                self.que.push(e.to);
+            }
+        }
+    }
+    fn dfs(&mut self, v: usize, up: Cap) -> Cap {
+        if v == self.s {
+            return up;
+        }
+        let mut res = Cap::zero();
+        let level_v = self.level[v];
+        for i in self.iter[v]..self.graph.g[v].len() {
+            self.iter[v] = i;
+            let &_Edge {
+                to: e_to,
+                rev: e_rev,
+                ..
+            } = &self.graph.g[v][i];
+            if level_v <= self.level[e_to] || self.graph.g[e_to][e_rev].cap == Cap::zero() {
+                continue;
+            }
+            let d = self.dfs(e_to, min(up - res, self.graph.g[e_to][e_rev].cap));
+            if d <= Cap::zero() {
+                continue;
+            }
+            self.graph.g[v][i].cap += d;
+            self.graph.g[e_to][e_rev].cap -= d;
+            res += d;
+            if res == up {
+                break;
+            }
+        }
```

Comment:
```suggestion
        }
        self.iter[v] = self.graph.g[v].len();
```
This is an important bit of the algorithm that we should not omit. Alternatively we can do
```
self.level[v] = self.graph.g.len();
```

Reply:
You're right.  It is difficult to translate C++ `for`-statements with side effects...